File: Project4/PythonTestingGrounds/2023-11-08_WildTree_Huffman_delta_IntegerCompression_BinaryWrite/tree_class.py
#! /usr/bin/python3
import array

class NodeClass:
	node_count = 0
	index_count = 0
	distinct_count = 0
	pop_count = 0
	push_count = 0
	temp1 = 0
	temp2 = 0
	temp3 = 0

	scan_stack = []

	file_ptr = None
	print_or_write = None

	def __init__(self):
		self.subnode = [None] * 27		# 0 to 26
		self.subnode[0] = []			# The 0th element has occurrence index count
		NodeClass.node_count += 1

	def addElement(self, word):
		# if the word has been encoded add in this sequence, add the index tag now.
		if len(word) == 0:
			NodeClass.index_count += 1
			# Keep track of distincts
			if len(self.subnode[0]) == 0:
				NodeClass.distinct_count += 1

			self.subnode[0].append(NodeClass.index_count)
			return

		# If there is still some of the word to be enqueued ...
		key = ord(word[0]) & 0x1F
		
		# if such a node does not exist, make one
		if self.subnode[key] is None:
			self.subnode[key] = NodeClass()

		# Keep enqueuing the word
		self.subnode[key].addElement(word[1:])

	# returns the node that could be the word or prefix.
	def lookup(self, word):
		# if the word is found at this node, return this node
		if len(word) == 0:
			return self
		
		# If not yet found ...
		key = ord(word[0]) & 0x1F

		# if such a subnode does not exist, return
		if self.subnode[key] == None:
			return None

		# if such a subnode exists, but is not terminal, look deeper still
		return self.subnode[key].lookup(word[1:])

	def searchElement(self, word):
		returned_node = self.lookup(word)

		if returned_node is None:
			print("Word not found")
		else:
			if len(returned_node.subnode[0]) != 0:
				print("Word found at indices:")
				# iterate over indices
				for idx in returned_node.subnode[0]:
					print(idx, end=", ")
				# print new line
				print()
			else:
				print(f"{word} is a prefix.")

		return returned_node

	def traverseNode(self):
		if len(self.subnode[0]) > 0:
			NodeClass.print_or_write(self.subnode[0])
			NodeClass.pop_count = 0
			NodeClass.push_count = 0
		
		# go deeper
		for idx in range(1, 27):
			if self.subnode[idx] != None:
				NodeClass.scan_stack.append(chr(96+idx))
				NodeClass.push_count += 1
				self.subnode[idx].traverseNode()
				NodeClass.scan_stack.pop()
				NodeClass.pop_count += 1

		# return void
		return

	####################### PRINTING SUBROUTINES #########################
	def print_details(idx_list):
		print("<", NodeClass.pop_count, ">", sep="")
		# Print this one
		for character in NodeClass.scan_stack:
			print(character, end="")
		
		print(": ", end="")

		for idx in idx_list:
			print(idx, end=", ")

		print()

	def write_details(idx_list):
		string_to_write  = "<" + str(NodeClass.pop_count)
		string_to_write += "".join(str(x) for x in NodeClass.scan_stack[len(NodeClass.scan_stack) - NodeClass.push_count:])
		string_to_write += "|"
		NodeClass.file_ptr.write(string_to_write.encode())

		# make a delta compressed idx_list
		compressed_idx_list = []

		for idx in range(1, len(idx_list)):
			compressed_idx_list.append(idx_list[idx] - idx_list[idx-1])

		# Write a short to show the number of elements
		data  = array.array("H", [len(idx_list)]).tobytes()

		# Write an int to show the original offset
		data += array.array("I", [idx_list[0]]).tobytes()

		# Write an array of shorts for deltas
		if len(compressed_idx_list) > 0:
			for elem in compressed_idx_list:
				if elem <= 2**14-1:					# Can put in 14 bits? [2'b10, 14'hXXXX]
					NodeClass.temp1 += 1
					data += array.array("H", [elem | 0x8000]).tobytes()	# elem is already < 2**14-1 so no need for & 0x3FFF
				elif elem <= 2**23-1:			# Can put in 23 bits? [1'b0, 23'hXXXXXX]
					NodeClass.temp2 += 1
					data += array.array("B", [elem >> 16]).tobytes()
					data += array.array("H", [elem & 0xFFFF]).tobytes()
				else:								# Fit in 30 bits. [2'b11, 30'hXXXXXXXX]
					NodeClass.temp3 += 1
					data += array.array("I", [elem | 0xC0000000]).tobytes()

		NodeClass.file_ptr.write(data)

		return

		

"""		USAGE RUBRIK

motherNode = NodeClass()

motherNode.addElement('asad')
motherNode.addElement('naveed')
motherNode.addElement('asad')
motherNode.addElement('naveeda')
motherNode.addElement('naveed')
motherNode.addElement('asad')

print("Node Count: ", NodeClass.node_count)

print("\n\nLOOK UP ...")
motherNode.searchElement('asad')
"""

File: Project4/CPP_Code/2023-09-11_PythonEquivalent_NoMultiThreading_NoSIMD/tree_class.cpp
#include "tree_class.h"

int treenode::index_count = 0;
int treenode::distinct_count = 0;
int treenode::push_count = 0;
int treenode::pop_count = 0;

int treenode::temp1=0, treenode::temp2=0, treenode::temp3=0;

std::ofstream treenode::fstream_compressed;
std::string treenode::scan_stack;

char *treenode::compressed_sequence = NULL;
unsigned treenode::c_idx = 0;

funcptr_t *treenode::print_or_write = NULL;

treenode::treenode() {
	// initialize subnodes with NUKK
	for(int i=0; i<27; i++)
		this->subnode[i] = NULL;

	return;
}

void treenode::addElement(std::string word) {
	static char key;

	if(word.length() == 0) {					// This must be the final node
		treenode::index_count++;
		// Keep track of distincts. Allocate more space to vector in anticipation
		if(this->idx_list.empty()) {
			treenode::distinct_count++;
			this->idx_list.reserve(256);
		}

		// add the index to the vector
		this->idx_list.push_back(index_count);

		return;
	}

	// For non-final nodes.
	key = word[0] & 0x1F;

	// If the subnode does not exist, create it
	if(this->subnode[key] == NULL)
		this->subnode[key] = new treenode();

	this->subnode[key]->addElement(word.substr(1));
}

treenode *treenode::createNodesFromWord(std::string word) {
	static char key;
	// if the word has been encoded add in this sequence, add the index tag now.
	if(word.length() == 0)
		return this;

	// If there is still some of the word to be enqueued ...
	key = word[0] & 0x1F;
	
	// If the subnode does not exist, create it
	if(this->subnode[key] == NULL)
		this->subnode[key] = new treenode();

	// Keep enqueuing the word
	return this->subnode[key]->createNodesFromWord(word.substr(1));
}

treenode *treenode::lookup(std::string word) {
	static char key;
	
	if(word.length() == 0)			// if node match is complete
		return this;

	// Keep looking deeper
	key = word[0] & 0x1F;

	if(this->subnode[key] == NULL)
		return NULL;

	return this->subnode[key]->lookup(word.substr(1));
}

treenode *treenode::searchElement(std::string word) {
	treenode *returned_node = this->lookup(word);

	if(returned_node == NULL)
		std::cout << "Word not found...";
	else {
		if(!returned_node->idx_list.empty()) {		// if not empty
			std::cout << "Word found at indices:\n";

			for(auto elem : returned_node->idx_list) 
  				std::cout << elem << ", ";
		}
		else {
			std::cout << word << " is a prefix.";
		}
	}

	std::cout << std::endl;

	return returned_node;
}

void treenode::traverseNode() {
	if(! this->idx_list.empty())
		print_or_write(this->idx_list);

	// go deeper
	for(int i=0; i<27; i++) {
		if(this->subnode[i] != NULL) {
			scan_stack.push_back(char(96+i));
			this->subnode[i]->traverseNode();
			scan_stack.pop_back();
		}
	}

	return;
}

void treenode::print_details(const std::vector<unsigned int> &idx_list) {
	std::cout << scan_stack << ": ";

	for (auto elem : idx_list) {
		std::cout << elem << ", ";
	}

	std::cout << "\n";

	return;
}

// format [1byte seqLength, N byte seq, 2 byte idx_list_length, 4 byte first idx, N byte compressed delta stream]
void treenode::write_details(const std::vector<unsigned int> &idx_list) {
	static unsigned len_var, delta;

	len_var = scan_stack.length();

	for(int i=0; i<len_var; i++)					// N byte sequence
		compressed_sequence[c_idx++] = scan_stack[i];		

	len_var = idx_list.size();	

	// 2 bytes for idx count. MSnibble is indicator for idx start
	*((unsigned short *) &compressed_sequence[c_idx]) = (unsigned short) (len_var | 0xF000);
	c_idx += 2;

	*((unsigned int *) &compressed_sequence[c_idx]) = (unsigned int) idx_list[0];
	c_idx += 4;

	for(int i=1; i<len_var; i++) {
		delta = idx_list[i] -  idx_list[i-1];

		// huffman coding
		if((delta & 0xFFFFC000) == 0) {			// Can put in 14 bits? [2'b10, 14'hXXXX]
			temp1++;
			*((unsigned short *) &compressed_sequence[c_idx]) = (unsigned short) (delta | 0x8000);
			c_idx += 2;
		}
		else if((delta & 0xFF800000) == 0) {		// Can put in 23 bits? [1'b0, 23'hXXXXXX]
			temp2++;
			*((unsigned int *) &compressed_sequence[c_idx]) = delta;
			c_idx += 3;
		}
		else {								// Fit in 30 bits. [2'b11, 30'hXXXXXXXX]
			temp3++;
			*((unsigned int *) &compressed_sequence[c_idx]) = delta | 0xC0000000;
			c_idx += 4;
		}
	}

	return;
}
